Sources/SwiftLexicalLookup/ScopeImplementations.swift

```diff
@@ -96,28 +96,80 @@ import SwiftSyntax
   /// - for `memberBlockUpToLastDecl` - a, b, c, d
   /// - for `memberBlock` - a, b, c, d, e, f
   /// - for `codeBlock` - a
-  @_spi(Experimental) public func lookup(
-    for name: String?,
+  @_spi(Experimental) public func _lookup(
+    for identifier: Identifier?,
     at syntax: SyntaxProtocol,
-    with config: LookupConfig
+    with config: LookupConfig,
+    state: LookupState
   ) -> [LookupResult] {
-    let names = introducedNames(using: config.fileScopeHandling)
-      .filter { introducedName in
-        introducedName.isAccessible(at: syntax) && (name == nil || introducedName.refersTo(name!))
+    switch config.fileScopeHandling {
+    case .memberBlock:
+      let names = introducedNames(using: .memberBlock)
+        .filter { lookupName in
+          checkName(identifier, refersTo: lookupName, at: syntax)
+        }
+
+      return names.isEmpty ? [] : [.fromFileScope(self, withNames: names)]
+    case .memberBlockUpToLastDecl:
+      var members = [LookupName]()
+      var sequentialItems = [CodeBlockItemSyntax]()
+      var encounteredNonDeclaration = false
+
+      for codeBlockItem in statements {
```

Comment:
I think I'm a little confused how should we treat the file scope. When I run this code snippet in a playground
```Swift
class Foo {
  func bar() {
    x.test()
  }
}

let x: X = X()

class X {
  func test() {
    print("It works")
  }
}

Foo().bar()
```
The result is:
```
It works
```
But when I move `Foo().bar()` to before `let x`:
```Swift
class Foo {
  func bar() {
    x.test()
  }
}

Foo().bar()

let x: X = X()

class X {
  func test() {
    print("It works")
  }
}
```
The code compiles (so x is already visible inside `Foo`), but we get `EXC_BAD_ACCESS`. 
Lastly, when we run
```Swift
class Foo {
  func bar() {
    print(x, y)
  }
}

let x: Int? = 1

Foo().bar()

let y: Int? = 2
```
The output is:
```
Optional(1) Optional(0)
```
So does the compiler hoist variables defined after the first non-declaration to the top of the file and default some of their values to some particular ones? Should we maybe represent this playground behavior with some kind of result flag saying "this variable hasn't been initialized yet"? From what I've seen here, the name availability should be similar to the one of member scope (but then also not really, because we also have to somehow handle `guard`). I'm curious what do you think about this.

Reply:
The name lookup rules at file scope are a bit... unfortunate, and you're seeing the effects of that in your example. Long-term, we'd like to fix the mistakes here.

I don't think we should try to determine "hasn't been initialized yet" here in this library. It's something that should be done in a later pass in the compiler that can reason about control flow more effectively.

For this library, we should try to match the existing lookup behavior, and document what it's doing. When it comes time to revisit this behavior (if ever), it'll be a lot easier to fix here than in the C++ implementation of unqualified lookup.

Reply:
> I'm a little confused myself as to why we're gathering declarations up to the first non-declaration, then gathering everything else as code block items, because I don't consider the first non-declaration to be special in Swift's lookup rules.

Does this mean that we should remove the `memberBlockUpToLastDecl` lookup kind?

Reply:
Maybe? There's definitely a difference here for the source file scope when top-level code is involved, and we need a way to model it. I think we should leave this in place for now and we can rename/revise as we learn more.